<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NACHO">
<title>NACHO • NACHO</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="NACHO">
<meta property="og:description" content="NACHO">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NACHO</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.0.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/NACHO.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/NACHO-analysis.html">NACHO Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mcanouil/NACHO/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>NACHO</h1>
            <h3 data-toc-skip class="subtitle">A NAnostring quality
Control dasHbOard</h3>
                        <h4 data-toc-skip class="author">Mickaël
Canouil, Ph.D., Gerard A. Bouland and Roderick C. Slieker, Ph.D.</h4>
            
            <h4 data-toc-skip class="date">January 12, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mcanouil/NACHO/blob/HEAD/vignettes/NACHO.Rmd" class="external-link"><code>vignettes/NACHO.Rmd</code></a></small>
      <div class="d-none name"><code>NACHO.Rmd</code></div>
    </div>

    
    
<p><img src="nacho_hex.png" width="150px" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install NACHO from CRAN:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"NACHO"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or the the development version from GitHub:</span></span>
<span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mcanouil/NACHO"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load NACHO</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mcanouil/NACHO/" class="external-link">NACHO</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><em>NACHO</em> (<strong>NA</strong>nostring quality
<strong>C</strong>ontrol das<strong>H</strong>b<strong>O</strong>ard) is
developed for NanoString nCounter data.<br>
NanoString nCounter data is a messenger-RNA/micro-RNA (mRNA/miRNA)
expression assay and works with fluorescent barcodes.<br>
Each barcode is assigned a mRNA/miRNA, which can be counted after
bonding with its target.<br>
As a result each count of a specific barcode represents the presence of
its target mRNA/miRNA.</p>
<p><em>NACHO</em> is able to load, visualise and normalise the exported
NanoString nCounter data and facilitates the user in performing a
quality control.<br><em>NACHO</em> does this by visualising quality control metrics,
expression of control genes, principal components and sample specific
size factors in an interactive web application.</p>
<p>With the use of two functions, RCC files are summarised and
visualised, namely: <code><a href="../reference/load_rcc.html">load_rcc()</a></code> and
<code><a href="../reference/visualise.html">visualise()</a></code>.</p>
<ul>
<li>The <code><a href="../reference/load_rcc.html">load_rcc()</a></code> function is used to preprocess the
data.</li>
<li>The <code><a href="../reference/visualise.html">visualise()</a></code> function initiates a <a href="https://shiny.posit.co/" class="external-link">Shiny-based dashboard</a> that visualises
all relevant QC plots.</li>
</ul>
<p><em>NACHO</em> also includes a function <code><a href="../reference/normalise.html">normalise()</a></code>,
which (re)calculates sample specific size factors and normalises the
data.</p>
<ul>
<li>The <code><a href="../reference/normalise.html">normalise()</a></code> function creates a list in which your
settings, the raw counts and normalised counts are stored.</li>
</ul>
<p>In addition (since v0.6.0) <em>NACHO</em> includes two (three)
additional functions:</p>
<ul>
<li>The <code><a href="../reference/render.html">render()</a></code> function renders a full quality-control
report (HTML) based on the results of a call to <code><a href="../reference/load_rcc.html">load_rcc()</a></code>
or <code><a href="../reference/normalise.html">normalise()</a></code> (using <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> in a Rmarkdown
chunk).</li>
<li>The <code><a href="../reference/autoplot.html">autoplot()</a></code> function draws any quality-control
metrics from <code><a href="../reference/visualise.html">visualise()</a></code> and <code><a href="../reference/render.html">render()</a></code>.</li>
</ul>
<p>For more <code><a href="../articles/NACHO.html">vignette("NACHO")</a></code> and
<code><a href="../articles/NACHO-analysis.html">vignette("NACHO-analysis")</a></code>.</p>
<p>
Canouil M, Bouland GA, Bonnefond A, Froguel P, Hart L, Slieker R (2019).
“NACHO: an R package for quality control of NanoString nCounter data.”
<em>Bioinformatics</em>. ISSN 1367-4803,
<a href="https://doi.org/10.1093/bioinformatics/btz647" class="external-link">doi:10.1093/bioinformatics/btz647</a>.
</p>
<pre><code>@Article{,
  title = {{NACHO}: an {R} package for quality control of {NanoString} {nCounter} data},
  author = {Mickaël Canouil and Gerard A. Bouland and Amélie Bonnefond and Philippe Froguel and Leen Hart and Roderick Slieker},
  journal = {Bioinformatics},
  address = {Oxford, England},
  year = {2019},
  month = {aug},
  issn = {1367-4803},
  doi = {10.1093/bioinformatics/btz647},
}</code></pre>
</div>
<div class="section level2">
<h2 id="an-example">An example<a class="anchor" aria-label="anchor" href="#an-example"></a>
</h2>
<p>To display the usage and utility of <em>NACHO</em>, we show three
examples in which the above mentioned functions are used and the results
are briefly examined.</p>
<p><em>NACHO</em> comes with presummarised data and in the first example
we use this dataset to call the interactive web application using
<code><a href="../reference/visualise.html">visualise()</a></code>.<br>
In the second example, we show the process of going from raw RCC files
to visualisations with a dataset queried from <strong>GEO</strong> using
<code>GEOquery</code>.<br>
In the third example, we use the summarised dataset from the second
example to calculate the sample specific size factors using
<code><a href="../reference/normalise.html">normalise()</a></code> and its added functionality to predict
housekeeping genes.</p>
<p>Besides creating interactive visualisations, <em>NACHO</em> also
identifies poorly performing samples which can be seen under the Outlier
Table tab in the interactive web application.<br>
While calling <code><a href="../reference/normalise.html">normalise()</a></code>, the user has the possibility to
remove these outliers before size factor calculation.</p>
<div class="section level3">
<h3 id="get-nanostring-ncounter-data">Get NanoString nCounter data<a class="anchor" aria-label="anchor" href="#get-nanostring-ncounter-data"></a>
</h3>
<div class="section level4">
<h4 id="presummarised-data-from-nacho">Presummarised data from <em>NACHO</em><a class="anchor" aria-label="anchor" href="#presummarised-data-from-nacho"></a>
</h4>
<p>This example shows how to use summarised data to call the interactive
web application.<br>
The raw data used is from a study of <span class="citation">Liu et al.
(<a href="#ref-liu_pam50_2016" role="doc-biblioref">2016</a>)</span> and
was acquired from the NCBI GEO public database <span class="citation">(<a href="#ref-barrett_ncbi_2013" role="doc-biblioref">Barrett et al. 2013</a>)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mcanouil/NACHO/" class="external-link">NACHO</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GSE74821</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/visualise.html">visualise</a></span><span class="op">(</span><span class="va">GSE74821</span><span class="op">)</span></span></code></pre></div>
<p><img src="README-visualise.png" width="650px" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="raw-data-from-geo">Raw data from GEO<a class="anchor" aria-label="anchor" href="#raw-data-from-geo"></a>
</h4>
<p>Numerous NanoString nCounter datasets are available from GEO <span class="citation">(<a href="#ref-barrett_ncbi_2013" role="doc-biblioref">Barrett et al. 2013</a>)</span>.<br>
In this example, we use a mRNA dataset from the study of <span class="citation">Bruce et al. (<a href="#ref-bruce_identification_2015" role="doc-biblioref">2015</a>)</span> with the GEO accession number:
<strong>GSE70970</strong>. The data is extracted and prepared using the
following code.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/seandavi/GEOquery" class="external-link">GEOquery</a></span><span class="op">)</span></span>
<span><span class="co"># Download data</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://seandavi.github.io/GEOquery/reference/getGEO.html" class="external-link">getGEO</a></span><span class="op">(</span><span class="st">"GSE70970"</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://seandavi.github.io/GEOquery/reference/getGEOSuppFiles.html" class="external-link">getGEOSuppFiles</a></span><span class="op">(</span>GEO <span class="op">=</span> <span class="st">"GSE70970"</span>, baseDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Unzip data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span></span>
<span>  tarfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"GSE70970"</span>, <span class="st">"GSE70970_RAW.tar"</span><span class="op">)</span>,</span>
<span>  exdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"GSE70970"</span>, <span class="st">"Data"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Get phenotypes and add IDs</span></span>
<span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">phenoData</a></span><span class="op">(</span><span class="va">gse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">targets</span><span class="op">$</span><span class="va">IDFILE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"GSE70970"</span>, <span class="st">"Data"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                 IDFILE                title geo_accession</span></span>
<span><span class="co">## GSM1824143   GSM1824143_NPC-T-1.RCC.gz   NPC-Training Set-1    GSM1824143</span></span>
<span><span class="co">## GSM1824144  GSM1824144_NPC-T-10.RCC.gz  NPC-Training Set-10    GSM1824144</span></span>
<span><span class="co">## GSM1824145 GSM1824145_NPC-T-100.RCC.gz NPC-Training Set-100    GSM1824145</span></span>
<span><span class="co">## GSM1824146 GSM1824146_NPC-T-101.RCC.gz NPC-Training Set-101    GSM1824146</span></span>
<span><span class="co">## GSM1824147 GSM1824147_NPC-T-102.RCC.gz NPC-Training Set-102    GSM1824147</span></span>
<span><span class="co">##                           status submission_date last_update_date type</span></span>
<span><span class="co">## GSM1824143 Public on Jul 17 2015     Jul 15 2015      Jul 20 2015  RNA</span></span>
<span><span class="co">## GSM1824144 Public on Jul 17 2015     Jul 15 2015      Jul 20 2015  RNA</span></span>
<span><span class="co">## GSM1824145 Public on Jul 17 2015     Jul 15 2015      Jul 20 2015  RNA</span></span>
<span><span class="co">## GSM1824146 Public on Jul 17 2015     Jul 15 2015      Jul 20 2015  RNA</span></span>
<span><span class="co">## GSM1824147 Public on Jul 17 2015     Jul 15 2015      Jul 20 2015  RNA</span></span>
<span><span class="co">##            channel_count                      source_name_ch1 organism_ch1</span></span>
<span><span class="co">## GSM1824143             1 FFPE Nasopharyngeal Carcinoma Biopsy Homo sapiens</span></span>
<span><span class="co">## GSM1824144             1 FFPE Nasopharyngeal Carcinoma Biopsy Homo sapiens</span></span>
<span><span class="co">## GSM1824145             1 FFPE Nasopharyngeal Carcinoma Biopsy Homo sapiens</span></span>
<span><span class="co">## GSM1824146             1 FFPE Nasopharyngeal Carcinoma Biopsy Homo sapiens</span></span>
<span><span class="co">## GSM1824147             1 FFPE Nasopharyngeal Carcinoma Biopsy Homo sapiens</span></span>
<span><span class="co">##            characteristics_ch1     characteristics_ch1.1 characteristics_ch1.2</span></span>
<span><span class="co">## GSM1824143    distant.event: 0 distant.time: 5.095140315     survival.event: 0</span></span>
<span><span class="co">## GSM1824144    distant.event: 0 distant.time: 7.471594798     survival.event: 0</span></span>
<span><span class="co">## GSM1824145    distant.event: 0   distant.time: 4.8678987     survival.event: 0</span></span>
<span><span class="co">## GSM1824146    distant.event: 0 distant.time: 3.835728953     survival.event: 0</span></span>
<span><span class="co">## GSM1824147    distant.event: 0 distant.time: 4.292950034     survival.event: 0</span></span>
<span><span class="co">##                 characteristics_ch1.3 characteristics_ch1.4</span></span>
<span><span class="co">## GSM1824143 survival.time: 5.095140315        local.event: 1</span></span>
<span><span class="co">## GSM1824144 survival.time: 7.471594798        local.event: 1</span></span>
<span><span class="co">## GSM1824145   survival.time: 4.8678987        local.event: 0</span></span>
<span><span class="co">## GSM1824146 survival.time: 3.835728953        local.event: 0</span></span>
<span><span class="co">## GSM1824147 survival.time: 4.292950034        local.event: 1</span></span>
<span><span class="co">##              characteristics_ch1.5            characteristics_ch1.6</span></span>
<span><span class="co">## GSM1824143 local.time: 5.075975359 local.regional.time: 5.075975359</span></span>
<span><span class="co">## GSM1824144 local.time: 1.204654346 local.regional.time: 1.204654346</span></span>
<span><span class="co">## GSM1824145   local.time: 4.8678987   local.regional.time: 4.8678987</span></span>
<span><span class="co">## GSM1824146 local.time: 3.835728953 local.regional.time: 3.835728953</span></span>
<span><span class="co">## GSM1824147 local.time: 2.828199863 local.regional.time: 2.828199863</span></span>
<span><span class="co">##              characteristics_ch1.7   characteristics_ch1.8</span></span>
<span><span class="co">## GSM1824143 local.regional.event: 1 nodal.time: 5.095140315</span></span>
<span><span class="co">## GSM1824144 local.regional.event: 1 nodal.time: 7.471594798</span></span>
<span><span class="co">## GSM1824145 local.regional.event: 0   nodal.time: 4.8678987</span></span>
<span><span class="co">## GSM1824146 local.regional.event: 0 nodal.time: 3.835728953</span></span>
<span><span class="co">## GSM1824147 local.regional.event: 1 nodal.time: 4.292950034</span></span>
<span><span class="co">##            characteristics_ch1.9 characteristics_ch1.10</span></span>
<span><span class="co">## GSM1824143        nodal.event: 0       disease.event: 1</span></span>
<span><span class="co">## GSM1824144        nodal.event: 0       disease.event: 1</span></span>
<span><span class="co">## GSM1824145        nodal.event: 0       disease.event: 0</span></span>
<span><span class="co">## GSM1824146        nodal.event: 0       disease.event: 0</span></span>
<span><span class="co">## GSM1824147        nodal.event: 0       disease.event: 1</span></span>
<span><span class="co">##               characteristics_ch1.11 characteristics_ch1.12</span></span>
<span><span class="co">## GSM1824143 disease.time: 5.075975359                  t: T1</span></span>
<span><span class="co">## GSM1824144 disease.time: 1.204654346                  t: T1</span></span>
<span><span class="co">## GSM1824145   disease.time: 4.8678987                  t: T1</span></span>
<span><span class="co">## GSM1824146 disease.time: 3.835728953                  t: T3</span></span>
<span><span class="co">## GSM1824147 disease.time: 2.828199863                  t: T3</span></span>
<span><span class="co">##            characteristics_ch1.13 characteristics_ch1.14</span></span>
<span><span class="co">## GSM1824143                  n: N0       age: 45.97260274</span></span>
<span><span class="co">## GSM1824144                 n: N2c              age: 46.4</span></span>
<span><span class="co">## GSM1824145                  n: N1       age: 50.36438356</span></span>
<span><span class="co">## GSM1824146                  n: N0       age: 64.09041096</span></span>
<span><span class="co">## GSM1824147                  n: N2       age: 27.57808219</span></span>
<span><span class="co">##                    characteristics_ch1.15 characteristics_ch1.16</span></span>
<span><span class="co">## GSM1824143 disease.spec.time: 5.095140315  disease.spec.event: 0</span></span>
<span><span class="co">## GSM1824144 disease.spec.time: 7.471594798  disease.spec.event: 0</span></span>
<span><span class="co">## GSM1824145   disease.spec.time: 4.8678987  disease.spec.event: 0</span></span>
<span><span class="co">## GSM1824146 disease.spec.time: 3.835728953  disease.spec.event: 0</span></span>
<span><span class="co">## GSM1824147 disease.spec.time: 4.292950034  disease.spec.event: 0</span></span>
<span><span class="co">##            characteristics_ch1.17 characteristics_ch1.18 characteristics_ch1.19</span></span>
<span><span class="co">## GSM1824143           gender: Male               chemo: 0               bin.t: 0</span></span>
<span><span class="co">## GSM1824144           gender: Male               chemo: 1               bin.t: 0</span></span>
<span><span class="co">## GSM1824145           gender: Male               chemo: 0               bin.t: 0</span></span>
<span><span class="co">## GSM1824146         gender: Female               chemo: 1               bin.t: 1</span></span>
<span><span class="co">## GSM1824147           gender: Male               chemo: 1               bin.t: 1</span></span>
<span><span class="co">##               treatment_protocol_ch1 growth_protocol_ch1 molecule_ch1</span></span>
<span><span class="co">## GSM1824143 Xylene de-paraffinization                  NA    total RNA</span></span>
<span><span class="co">## GSM1824144 Xylene de-paraffinization                  NA    total RNA</span></span>
<span><span class="co">## GSM1824145 Xylene de-paraffinization                  NA    total RNA</span></span>
<span><span class="co">## GSM1824146 Xylene de-paraffinization                  NA    total RNA</span></span>
<span><span class="co">## GSM1824147 Xylene de-paraffinization                  NA    total RNA</span></span>
<span><span class="co">##                                                           extract_protocol_ch1</span></span>
<span><span class="co">## GSM1824143 Recover All Total Nucleic Acid Isolation Kit for FFPE (Ambion, Inc)</span></span>
<span><span class="co">## GSM1824144 Recover All Total Nucleic Acid Isolation Kit for FFPE (Ambion, Inc)</span></span>
<span><span class="co">## GSM1824145 Recover All Total Nucleic Acid Isolation Kit for FFPE (Ambion, Inc)</span></span>
<span><span class="co">## GSM1824146 Recover All Total Nucleic Acid Isolation Kit for FFPE (Ambion, Inc)</span></span>
<span><span class="co">## GSM1824147 Recover All Total Nucleic Acid Isolation Kit for FFPE (Ambion, Inc)</span></span>
<span><span class="co">##                                 label_ch1                  label_protocol_ch1</span></span>
<span><span class="co">## GSM1824143 Nanostring fluorophore barcode Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824144 Nanostring fluorophore barcode Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824145 Nanostring fluorophore barcode Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824146 Nanostring fluorophore barcode Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824147 Nanostring fluorophore barcode Nanostring nCounter Analysis System</span></span>
<span><span class="co">##            taxid_ch1                        hyb_protocol</span></span>
<span><span class="co">## GSM1824143      9606 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824144      9606 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824145      9606 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824146      9606 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824147      9606 Nanostring nCounter Analysis System</span></span>
<span><span class="co">##                                  scan_protocol</span></span>
<span><span class="co">## GSM1824143 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824144 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824145 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824146 Nanostring nCounter Analysis System</span></span>
<span><span class="co">## GSM1824147 Nanostring nCounter Analysis System</span></span>
<span><span class="co">##                                                                                           data_processing</span></span>
<span><span class="co">## GSM1824143 variance stabilization and normalization (vsn) called through the NanoStringNorm package for R</span></span>
<span><span class="co">## GSM1824144 variance stabilization and normalization (vsn) called through the NanoStringNorm package for R</span></span>
<span><span class="co">## GSM1824145 variance stabilization and normalization (vsn) called through the NanoStringNorm package for R</span></span>
<span><span class="co">## GSM1824146 variance stabilization and normalization (vsn) called through the NanoStringNorm package for R</span></span>
<span><span class="co">## GSM1824147 variance stabilization and normalization (vsn) called through the NanoStringNorm package for R</span></span>
<span><span class="co">##            platform_id contact_name         contact_email</span></span>
<span><span class="co">## GSM1824143    GPL20699  Jeff,,Bruce jbruce@uhnresearch.ca</span></span>
<span><span class="co">## GSM1824144    GPL20699  Jeff,,Bruce jbruce@uhnresearch.ca</span></span>
<span><span class="co">## GSM1824145    GPL20699  Jeff,,Bruce jbruce@uhnresearch.ca</span></span>
<span><span class="co">## GSM1824146    GPL20699  Jeff,,Bruce jbruce@uhnresearch.ca</span></span>
<span><span class="co">## GSM1824147    GPL20699  Jeff,,Bruce jbruce@uhnresearch.ca</span></span>
<span><span class="co">##                          contact_institute contact_address contact_city</span></span>
<span><span class="co">## GSM1824143 Princess Margaret Cancer Centre 101 College St.      Toronto</span></span>
<span><span class="co">## GSM1824144 Princess Margaret Cancer Centre 101 College St.      Toronto</span></span>
<span><span class="co">## GSM1824145 Princess Margaret Cancer Centre 101 College St.      Toronto</span></span>
<span><span class="co">## GSM1824146 Princess Margaret Cancer Centre 101 College St.      Toronto</span></span>
<span><span class="co">## GSM1824147 Princess Margaret Cancer Centre 101 College St.      Toronto</span></span>
<span><span class="co">##            contact_state contact_zip/postal_code contact_country</span></span>
<span><span class="co">## GSM1824143       Ontario                  M5G1L7          Canada</span></span>
<span><span class="co">## GSM1824144       Ontario                  M5G1L7          Canada</span></span>
<span><span class="co">## GSM1824145       Ontario                  M5G1L7          Canada</span></span>
<span><span class="co">## GSM1824146       Ontario                  M5G1L7          Canada</span></span>
<span><span class="co">## GSM1824147       Ontario                  M5G1L7          Canada</span></span>
<span><span class="co">##                                                                                        supplementary_file</span></span>
<span><span class="co">## GSM1824143   ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1824nnn/GSM1824143/suppl/GSM1824143_NPC-T-1.RCC.gz</span></span>
<span><span class="co">## GSM1824144  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1824nnn/GSM1824144/suppl/GSM1824144_NPC-T-10.RCC.gz</span></span>
<span><span class="co">## GSM1824145 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1824nnn/GSM1824145/suppl/GSM1824145_NPC-T-100.RCC.gz</span></span>
<span><span class="co">## GSM1824146 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1824nnn/GSM1824146/suppl/GSM1824146_NPC-T-101.RCC.gz</span></span>
<span><span class="co">## GSM1824147 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1824nnn/GSM1824147/suppl/GSM1824147_NPC-T-102.RCC.gz</span></span>
<span><span class="co">##            data_row_count     age:ch1 bin.t:ch1 chemo:ch1 disease.event:ch1</span></span>
<span><span class="co">## GSM1824143            734 45.97260274         0         0                 1</span></span>
<span><span class="co">## GSM1824144            734        46.4         0         1                 1</span></span>
<span><span class="co">## GSM1824145            734 50.36438356         0         0                 0</span></span>
<span><span class="co">## GSM1824146            734 64.09041096         1         1                 0</span></span>
<span><span class="co">## GSM1824147            734 27.57808219         1         1                 1</span></span>
<span><span class="co">##            disease.spec.event:ch1 disease.spec.time:ch1 disease.time:ch1</span></span>
<span><span class="co">## GSM1824143                      0           5.095140315      5.075975359</span></span>
<span><span class="co">## GSM1824144                      0           7.471594798      1.204654346</span></span>
<span><span class="co">## GSM1824145                      0             4.8678987        4.8678987</span></span>
<span><span class="co">## GSM1824146                      0           3.835728953      3.835728953</span></span>
<span><span class="co">## GSM1824147                      0           4.292950034      2.828199863</span></span>
<span><span class="co">##            distant.event:ch1 distant.time:ch1 gender:ch1 local.event:ch1</span></span>
<span><span class="co">## GSM1824143                 0      5.095140315       Male               1</span></span>
<span><span class="co">## GSM1824144                 0      7.471594798       Male               1</span></span>
<span><span class="co">## GSM1824145                 0        4.8678987       Male               0</span></span>
<span><span class="co">## GSM1824146                 0      3.835728953     Female               0</span></span>
<span><span class="co">## GSM1824147                 0      4.292950034       Male               1</span></span>
<span><span class="co">##            local.regional.event:ch1 local.regional.time:ch1 local.time:ch1</span></span>
<span><span class="co">## GSM1824143                        1             5.075975359    5.075975359</span></span>
<span><span class="co">## GSM1824144                        1             1.204654346    1.204654346</span></span>
<span><span class="co">## GSM1824145                        0               4.8678987      4.8678987</span></span>
<span><span class="co">## GSM1824146                        0             3.835728953    3.835728953</span></span>
<span><span class="co">## GSM1824147                        1             2.828199863    2.828199863</span></span>
<span><span class="co">##            n:ch1 nodal.event:ch1 nodal.time:ch1 survival.event:ch1</span></span>
<span><span class="co">## GSM1824143    N0               0    5.095140315                  0</span></span>
<span><span class="co">## GSM1824144   N2c               0    7.471594798                  0</span></span>
<span><span class="co">## GSM1824145    N1               0      4.8678987                  0</span></span>
<span><span class="co">## GSM1824146    N0               0    3.835728953                  0</span></span>
<span><span class="co">## GSM1824147    N2               0    4.292950034                  0</span></span>
<span><span class="co">##            survival.time:ch1 t:ch1</span></span>
<span><span class="co">## GSM1824143       5.095140315    T1</span></span>
<span><span class="co">## GSM1824144       7.471594798    T1</span></span>
<span><span class="co">## GSM1824145         4.8678987    T1</span></span>
<span><span class="co">## GSM1824146       3.835728953    T3</span></span>
<span><span class="co">## GSM1824147       4.292950034    T3</span></span></code></pre>
<p>After we extracted the dataset to the
<code>/tmp/Rtmp8QDtI5/GSE70970/Data</code> directory, a
<code>Samplesheet.csv</code> containing a column with the exact names of
the files for each sample can be written or use as is.</p>
</div>
</div>
<div class="section level3">
<h3 id="the-load_rcc-function">The <code>load_rcc()</code> function<a class="anchor" aria-label="anchor" href="#the-load_rcc-function"></a>
</h3>
<p>The first argument requires the path to the directory containing the
RCC files, the second argument is the location of samplesheet followed
by third argument with the column name containing the exact names of the
files.<br>
The <code>housekeeping_genes</code> and
<code>normalisation_method</code> arguments respectively indicate which
housekeeping genes and normalisation method should be used.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE70970_sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_rcc.html">load_rcc</a></span><span class="op">(</span></span>
<span>  data_directory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"GSE70970"</span>, <span class="st">"Data"</span><span class="op">)</span>, <span class="co"># Where the data is</span></span>
<span>  ssheet_csv <span class="op">=</span> <span class="va">targets</span>, <span class="co"># The samplesheet</span></span>
<span>  id_colname <span class="op">=</span> <span class="st">"IDFILE"</span>, <span class="co"># Name of the column that contains the unique identfiers</span></span>
<span>  housekeeping_genes <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># Custom list of housekeeping genes</span></span>
<span>  housekeeping_predict <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># Whether or not to predict the housekeeping genes</span></span>
<span>  normalisation_method <span class="op">=</span> <span class="st">"GEO"</span>, <span class="co"># Geometric mean or GLM</span></span>
<span>  n_comp <span class="op">=</span> <span class="fl">5</span> <span class="co"># Number indicating how many principal components should be computed.</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## [NACHO] Importing RCC files.</span></span>
<span><span class="co">## Error in load_rcc(data_directory = file.path(tempdir(), "GSE70970", "Data"), : [NACHO] Multiple Nanostring file/software versions detected.</span></span>
<span><span class="co">##   Please provide a set of files with the same version.</span></span>
<span><span class="co">##   - FileVersion: '1.6', '1.6'</span></span>
<span><span class="co">##   - SoftwareVersion: '2.1.2.3', '2.1.1.0005'</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-visualise-function">The <code>visualise()</code> function<a class="anchor" aria-label="anchor" href="#the-visualise-function"></a>
</h3>
<p>When the summarisation is done, the summarised (or normalised) data
can be visualised using the <code><a href="../reference/visualise.html">visualise()</a></code> function as can be
seen in the following chunk of code.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualise.html">visualise</a></span><span class="op">(</span><span class="va">GSE70970_sum</span><span class="op">)</span></span></code></pre></div>
<p>The sidebar includes widgets to control quality-control thresholds.
These widgets differ according to the selected tab. Each sample in the
plots can be coloured based on either technical specifications which are
included in the RCC files or based on specifications of your own
choosing, though these specifications need to be included in the
samplesheet.</p>
</div>
<div class="section level3">
<h3 id="the-normalise-function">The <code>normalise()</code> function<a class="anchor" aria-label="anchor" href="#the-normalise-function"></a>
</h3>
<p><em>NACHO</em> allows the discovery of housekeeping genes within your
own dataset. <em>NACHO</em> finds the five best suitable housekeeping
genes, however, it is possible that one of these five genes might not be
suitable, which is why a subset of these discovered housekeeping genes
might work better in some cases. For this example, we use the
<strong>GSE70970</strong> dataset from the previous example. The
discovered housekeeping genes are saved in the result object as
<strong>predicted_housekeeping</strong>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">GSE70970_sum</span><span class="op">[[</span><span class="st">"housekeeping_genes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Error in eval(expr, envir, enclos): object 'GSE70970_sum' not found</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in eval(expr, envir, enclos): object 'GSE70970_sum' not found</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_housekeeping</span> <span class="op">&lt;-</span> <span class="va">GSE70970_sum</span><span class="op">[[</span><span class="st">"housekeeping_genes"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Error in eval(expr, envir, enclos): object 'GSE70970_sum' not found</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">my_housekeeping</span><span class="op">)</span></span>
<span><span class="co">## Error in eval(expr, envir, enclos): object 'my_housekeeping' not found</span></span></code></pre></div>
<p>The next step is the actual normalisation. The first argument
requires the summary which is created with the <code><a href="../reference/load_rcc.html">load_rcc()</a></code>
function. The second argument requires a vector of gene names. In this
case, it is a subset of the discovered housekeeping genes we just made.
With the third argument the user has the choice to remove the outliers.
Lastly, the normalisation method can be choosed.<br>
Here, the user has a choice between <code>"GLM"</code> or
<code>"GEO"</code>. The differences between normalisation methods are
nuanced, however, a preference for either method are use case
specific.<br>
In this example, <code>"GLM"</code> is used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GSE70970_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise.html">normalise</a></span><span class="op">(</span></span>
<span>  nacho_object <span class="op">=</span> <span class="va">GSE70970_sum</span>,</span>
<span>  housekeeping_genes <span class="op">=</span> <span class="va">my_housekeeping</span>,</span>
<span>  housekeeping_predict <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  housekeeping_norm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalisation_method <span class="op">=</span> <span class="st">"GEO"</span>,</span>
<span>  remove_outliers <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Error in eval(expr, envir, enclos): object 'GSE70970_sum' not found</span></span></code></pre></div>
<p><code><a href="../reference/normalise.html">normalise()</a></code> returns a <code>list</code> object (same as
<code><a href="../reference/load_rcc.html">load_rcc()</a></code>) with <code>raw_counts</code> and
<code>normalised_counts</code> slots filled with the raw and normalised
counts. Both counts are also in the <em>NACHO</em> data.frame.</p>
</div>
<div class="section level3">
<h3 id="the-autoplot-function">The <code>autoplot()</code> function<a class="anchor" aria-label="anchor" href="#the-autoplot-function"></a>
</h3>
<p>The <code><a href="../reference/autoplot.html">autoplot()</a></code> function provides an easy way to plot any
quality-control from the <code><a href="../reference/visualise.html">visualise()</a></code> function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/autoplot.html">autoplot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">GSE74821</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"BD"</span>,</span>
<span>  colour <span class="op">=</span> <span class="st">"CartridgeID"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The possible metrics (<code>x</code>) are:</p>
<ul>
<li>
<code>"BD"</code> (Binding Density)</li>
<li>
<code>"FoV"</code> (Imaging)</li>
<li>
<code>"PCL"</code> (Positive Control Linearity)</li>
<li>
<code>"LoD"</code> (Limit of Detection)</li>
<li>
<code>"Positive"</code> (Positive Controls)</li>
<li>
<code>"Negative"</code> (Negative Controls)</li>
<li>
<code>"Housekeeping"</code> (Housekeeping Genes)</li>
<li>
<code>"PN"</code> (Positive Controls vs. Negative Controls)</li>
<li>
<code>"ACBD"</code> (Average Counts vs. Binding Density)</li>
<li>
<code>"ACMC"</code> (Average Counts vs. Median Counts)</li>
<li>
<code>"PCA12"</code> (Principal Component 1 vs. 2)</li>
<li>
<code>"PCAi"</code> (Principal Component scree plot)</li>
<li>
<code>"PCA"</code> (Principal Components planes)</li>
<li>
<code>"PFNF"</code> (Positive Factor vs. Negative Factor)</li>
<li>
<code>"HF"</code> (Housekeeping Factor)</li>
<li>
<code>"NORM"</code> (Normalisation Factor)</li>
</ul>
<div class="section level4">
<h4 id="binding-density">Binding Density<a class="anchor" aria-label="anchor" href="#binding-density"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="imaging">Imaging<a class="anchor" aria-label="anchor" href="#imaging"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-2.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="positive-control-linearity">Positive Control Linearity<a class="anchor" aria-label="anchor" href="#positive-control-linearity"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-3.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="limit-of-detection">Limit of Detection<a class="anchor" aria-label="anchor" href="#limit-of-detection"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-4.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="positive-controls">Positive Controls<a class="anchor" aria-label="anchor" href="#positive-controls"></a>
</h4>
<pre><code><span><span class="co">## Warning: The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">colour</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span></span>
<span><span class="co">##   the data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span></span>
<span><span class="co">##   variable into a factor?</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-5.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="negative-controls">Negative Controls<a class="anchor" aria-label="anchor" href="#negative-controls"></a>
</h4>
<pre><code><span><span class="co">## Warning: The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">colour</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span></span>
<span><span class="co">##   the data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span></span>
<span><span class="co">##   variable into a factor?</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-6.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="housekeeping-genes">Housekeeping Genes<a class="anchor" aria-label="anchor" href="#housekeeping-genes"></a>
</h4>
<pre><code><span><span class="co">## Warning: The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">colour</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span></span>
<span><span class="co">##   the data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span></span>
<span><span class="co">##   variable into a factor?</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-7.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="positive-controls-vs--negative-controls">Positive Controls vs. Negative Controls<a class="anchor" aria-label="anchor" href="#positive-controls-vs--negative-controls"></a>
</h4>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-8.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="average-counts-vs--binding-density">Average Counts vs. Binding Density<a class="anchor" aria-label="anchor" href="#average-counts-vs--binding-density"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-9.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="average-counts-vs--median-counts">Average Counts vs. Median Counts<a class="anchor" aria-label="anchor" href="#average-counts-vs--median-counts"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-10.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="principal-component-1-vs--2">Principal Component 1 vs. 2<a class="anchor" aria-label="anchor" href="#principal-component-1-vs--2"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-11.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="principal-component-scree-plot">Principal Component scree plot<a class="anchor" aria-label="anchor" href="#principal-component-scree-plot"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-12.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="principal-components-planes">Principal Components planes<a class="anchor" aria-label="anchor" href="#principal-components-planes"></a>
</h4>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-13.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="positive-factor-vs--negative-factor">Positive Factor vs. Negative Factor<a class="anchor" aria-label="anchor" href="#positive-factor-vs--negative-factor"></a>
</h4>
<pre><code><span><span class="co">## Warning: Transformation introduced infinite values in continuous y-axis</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-14.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="housekeeping-factor">Housekeeping Factor<a class="anchor" aria-label="anchor" href="#housekeeping-factor"></a>
</h4>
<pre><code><span><span class="co">## Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">## Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">## Warning: Transformation introduced infinite values in continuous y-axis</span></span>
<span><span class="co">## Transformation introduced infinite values in continuous y-axis</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-15.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="normalisation-factor">Normalisation Factor<a class="anchor" aria-label="anchor" href="#normalisation-factor"></a>
</h4>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="NACHO_files/figure-html/unnamed-chunk-11-16.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="nacho-as-a-standalone-app">NACHO as a standalone app<a class="anchor" aria-label="anchor" href="#nacho-as-a-standalone-app"></a>
</h3>
<p><em>NACHO</em> is also available as a standalone app to be used in a
shiny server configuration. A convenience function <code><a href="../reference/deploy.html">deploy()</a></code>
is available to directly copy the <em>NACHO</em> app to the default
directory of a shiny server.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/deploy.html">deploy</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">"/srv/shiny-server"</span>, app_name <span class="op">=</span> <span class="st">"NACHO"</span><span class="op">)</span></span></code></pre></div>
<p>The app can also be run directly, without manually summarising and
normalising RCC files:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"app"</span>, package <span class="op">=</span> <span class="st">"NACHO"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="README-app.png" width="650px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="the-render-function">The <code>render()</code> function<a class="anchor" aria-label="anchor" href="#the-render-function"></a>
</h3>
<p>The <code><a href="../reference/render.html">render()</a></code> function renders a comprehensive HTML
report, using <code>print(..., echo = TRUE)</code>, which includes all
quality-control metrics and description of those metrics.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/render.html">render</a></span><span class="op">(</span></span>
<span>  nacho_object <span class="op">=</span> <span class="va">GSE74821</span>,</span>
<span>  colour <span class="op">=</span> <span class="st">"CartridgeID"</span>,</span>
<span>  output_file <span class="op">=</span> <span class="st">"NACHO_QC.html"</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"."</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clean <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The underneath function <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> can be used directly
within any Rmarkdown chunk, setting the parameter
<code>echo = TRUE</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">GSE74821</span>,</span>
<span>  colour <span class="op">=</span> <span class="st">"CartridgeID"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  echo <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  title_level <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="rcc-summary">RCC Summary<a class="anchor" aria-label="anchor" href="#rcc-summary"></a>
</h4>
<ul>
<li>Samples: 48</li>
<li>Endogenous: 50</li>
<li>Housekeeping: 8</li>
<li>Negative: 8</li>
<li>Positive: 6</li>
</ul>
</div>
<div class="section level4">
<h4 id="settings">Settings<a class="anchor" aria-label="anchor" href="#settings"></a>
</h4>
<ul>
<li><p>Predict housekeeping genes: FALSE</p></li>
<li><p>Normalise using housekeeping genes: TRUE</p></li>
<li><p>Housekeeping genes available: MRPL19, PSMC4, SF3A1, RPLP0, PUM1,
ACTB, TFRC and GUSB</p></li>
<li><p>Normalise using: GLM</p></li>
<li><p>Principal components to compute: 10</p></li>
<li>
<p>Remove outliers: FALSE</p>
<ul>
<li>Binding Density (BD) &lt; 0.1</li>
<li>Binding Density (BD) &gt; 2.25</li>
<li>Field of View (FoV) &lt; 75</li>
<li>Positive Control Linearity (PCL) &lt; 0.95</li>
<li>Limit of Detection (LoD) &lt; 2</li>
<li>Positive normalisation factor (Positive_factor) &lt; 0.25</li>
<li>Positive normalisation factor (Positive_factor) &gt; 4</li>
<li>Housekeeping normalisation factor (house_factor) &lt; 0.091</li>
<li>Housekeeping normalisation factor (house_factor) &gt; 11</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="qc-metrics">QC Metrics<a class="anchor" aria-label="anchor" href="#qc-metrics"></a>
</h4>
<div class="section level5">
<h5 id="binding-density-1">Binding Density<a class="anchor" aria-label="anchor" href="#binding-density-1"></a>
</h5>
<p>The imaging unit only counts the codes that are unambiguously
distinguishable.<br>
It simply will not count codes that overlap within an image.<br>
This provides increased confidence that the molecular counts you receive
are from truly recognisable codes.<br>
Under most conditions, forgoing the few barcodes that do overlap will
not impact your data.<br>
Too many overlapping codes in the image, however, will create a
condition called image saturation in which significant data loss could
occur (critical data loss from saturation is uncommon).</p>
<p>To determine the level of image saturation, the nCounter instrument
calculates the number of optical features per square micron for each
lane as it processes the images.<br>
This is called the <strong>Binding Density</strong>
(<strong>BD</strong>).<br>
The <strong>Binding Density</strong> is useful for determining whether
data collection has been compromised due to image saturation. The
acceptable range for <strong>Binding Density</strong> is:</p>
<ul>
<li>
<code>0.1 - 2.25</code> for
<strong>MAX</strong>/<strong>FLEX</strong> instruments</li>
<li>
<code>0.1 - 1.8</code> for <strong>SPRINT</strong> instruments</li>
</ul>
<p>Within these ranges, relatively few reporters on the slide surface
will overlap, enabling the instrument to accurately tabulate counts for
each reporter species.<br>
A <strong>Binding Density</strong> significantly greater than the upper
limit in either range is indicative of overlapping reporters on the
slide surface.<br>
The counts observed in lanes with a <strong>Binding Density</strong> at
this level may have had significant numbers of codes ignored, which
could potentially affect quantification and linearity of the assay.</p>
<p><img src="NACHO_files/figure-html/print-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="field-of-view-imaging">Field of View (Imaging)<a class="anchor" aria-label="anchor" href="#field-of-view-imaging"></a>
</h5>
<p>Each individual lane scanned on an nCounter system is divided into a
few hundred imaging sections, called Fields of View
(<strong>FOV</strong>), the exact number of which will depend on the
system being used (<em>i.e.</em>, <strong>MAX/FLEX</strong> or
<strong>SPRINT</strong>), and the scanner settings selected by the
user.<br>
The system images these <strong>FOV</strong>s separately, and sums the
barcode counts of all <strong>FOV</strong>s from a single lane to form
the final raw data count for each unique barcode target.<br>
Finally, the system reports the number of <strong>FOV</strong>s
successfully imaged as <strong>FOV</strong> Counted.</p>
<p>Significant discrepancy between the number of <strong>FOV</strong>
for which imaging was attempted (<strong>FOV Count</strong>) and for
which imaging was successful (<strong>FOV Counted</strong>) may indicate
an issue with imaging performance.<br>
Recommended percentage of registered FOVs (<em>i.e.</em>, <strong>FOV
Counted</strong> over <strong>FOV Count</strong>) is
<code>75 %</code>.</p>
<p><img src="NACHO_files/figure-html/print-2.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="positive-control-linearity-1">Positive Control Linearity<a class="anchor" aria-label="anchor" href="#positive-control-linearity-1"></a>
</h5>
<p>Six synthetic DNA control targets are included with every nCounter
Gene Expression assay.<br>
Their concentrations range linearly (in <em>codeset</em>) from
<code>128 fM</code> to <code>0.125 fM</code>, and they are referred to
as <strong>POS_A</strong> to <strong>POS_F</strong>, respectively.<br>
These <strong>Positive Controls</strong> are typically used to measure
the efficiency of the hybridization reaction, and their step-wise
concentrations also make them useful in checking the linearity
performance of the assay.</p>
<p>Since the known concentrations of the <strong>Positive
Controls</strong> increase in a linear fashion, the resulting counts
should, as well.</p>
<!--
Note that because **POS_F** has a known concentration of `0.125 fM`, which is considered below the limit of detection of the system, it should be excluded from this calculation (although you will see that **POS_F** counts are significantly higher than the negative control counts in most cases).
-->
<p><img src="NACHO_files/figure-html/print-3.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="limit-of-detection-1">Limit of Detection<a class="anchor" aria-label="anchor" href="#limit-of-detection-1"></a>
</h5>
<p>The limit of detection (<strong>LoD</strong>) is determined by
measuring the ability to detect <strong>POS_E</strong>, the
<code>0.5 fM</code> positive control probe, which corresponds to about
10,000 copies of this target within each sample tube.<br>
On a <strong>FLEX</strong>/<strong>MAX</strong> system, the standard
input of <code>100 ng</code> of total RNA will roughly correspond to
about 10,000 cell equivalents (assuming one cell contains
<code>10 pg</code> total RNA on average).<br>
An nCounter assay run on the <strong>FLEX</strong>/<strong>MAX</strong>
system should thus conservatively be able to detect roughly one
transcript copy per cell for each target (or 10,000 total transcript
copies).<br>
In most (codeset) assays, you will observe that even the
<strong>POS_F</strong> probe (equivalent to 0.25 copies per cell) is
detectable above background.</p>
<p><img src="NACHO_files/figure-html/print-4.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level4">
<h4 id="control-genes">Control Genes<a class="anchor" aria-label="anchor" href="#control-genes"></a>
</h4>
<div class="section level5">
<h5 id="positive">Positive<a class="anchor" aria-label="anchor" href="#positive"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-5.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="negative">Negative<a class="anchor" aria-label="anchor" href="#negative"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-6.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="housekeeping">Housekeeping<a class="anchor" aria-label="anchor" href="#housekeeping"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-7.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="control-probe-expression">Control Probe Expression<a class="anchor" aria-label="anchor" href="#control-probe-expression"></a>
</h5>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="NACHO_files/figure-html/print-8.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level4">
<h4 id="quality-control-visuals">Quality-Control Visuals<a class="anchor" aria-label="anchor" href="#quality-control-visuals"></a>
</h4>
<div class="section level5">
<h5 id="average-count-vs--binding-density">Average Count vs. Binding Density<a class="anchor" aria-label="anchor" href="#average-count-vs--binding-density"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-9.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="average-count-vs--median-count">Average Count vs. Median Count<a class="anchor" aria-label="anchor" href="#average-count-vs--median-count"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-10.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="principal-component">Principal Component<a class="anchor" aria-label="anchor" href="#principal-component"></a>
</h5>
<div class="section level6">
<h6 id="pc1-vs--pc2">PC1 vs. PC2<a class="anchor" aria-label="anchor" href="#pc1-vs--pc2"></a>
</h6>
<p><img src="NACHO_files/figure-html/print-11.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="factorial-planes">Factorial planes<a class="anchor" aria-label="anchor" href="#factorial-planes"></a>
</h6>
<p><img src="NACHO_files/figure-html/print-12.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="proportion-of-variance-explained">Proportion of Variance Explained<a class="anchor" aria-label="anchor" href="#proportion-of-variance-explained"></a>
</h6>
<p><img src="NACHO_files/figure-html/print-13.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level4">
<h4 id="normalisation">Normalisation<a class="anchor" aria-label="anchor" href="#normalisation"></a>
</h4>
<div class="section level5">
<h5 id="positive-factor-vs--background-threshold">Positive Factor vs. Background Threshold<a class="anchor" aria-label="anchor" href="#positive-factor-vs--background-threshold"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-14.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="housekeeping-factor-1">Housekeeping Factor<a class="anchor" aria-label="anchor" href="#housekeeping-factor-1"></a>
</h5>
<p><img src="NACHO_files/figure-html/print-15.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="normalisation-result">Normalisation Result<a class="anchor" aria-label="anchor" href="#normalisation-result"></a>
</h5>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="NACHO_files/figure-html/print-16.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-barrett_ncbi_2013" class="csl-entry">
Barrett, Tanya, Stephen E. Wilhite, Pierre Ledoux, Carlos Evangelista,
Irene F. Kim, Maxim Tomashevsky, Kimberly A. Marshall, et al. 2013.
<span>“<span>NCBI</span> <span>GEO</span>: Archive for Functional
Genomics Data Sets—Update.”</span> <em>Nucleic Acids Research</em> 41
(January): D991–95. <a href="https://doi.org/10.1093/nar/gks1193" class="external-link">https://doi.org/10.1093/nar/gks1193</a>.
</div>
<div id="ref-bruce_identification_2015" class="csl-entry">
Bruce, Jeff P., Angela B. Y. Hui, Wei Shi, Bayardo Perez-Ordonez, Ilan
Weinreb, Wei Xu, Benjamin Haibe-Kains, et al. 2015.
<span>“Identification of a <span class="nocase">microRNA</span>
Signature Associated with Risk of Distant Metastasis in Nasopharyngeal
Carcinoma.”</span> <em>Oncotarget</em> 6 (6): 4537–50. <a href="https://doi.org/10.18632/oncotarget.3005" class="external-link">https://doi.org/10.18632/oncotarget.3005</a>.
</div>
<div id="ref-liu_pam50_2016" class="csl-entry">
Liu, Minetta C., Brandelyn N. Pitcher, Elaine R. Mardis, Sherri R.
Davies, Paula N. Friedman, Jacqueline E. Snider, Tammi L. Vickery, et
al. 2016. <span>“<span>PAM</span>50 Gene Signatures and Breast Cancer
Prognosis with Adjuvant Anthracycline- and Taxane-Based Chemotherapy:
Correlative Analysis of C9741 (Alliance).”</span> <em>Npj Breast
Cancer</em> 2 (January): 15023. <a href="https://doi.org/10.1038/npjbcancer.2015.23" class="external-link">https://doi.org/10.1038/npjbcancer.2015.23</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://mickael.canouil.fr" class="external-link">Mickaël Canouil</a>, Roderick Slieker, Gerard Bouland.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
